File: 3rd_part_RO_lab_algorithms_CLI/python/ex5_grid_paths.py
def grid_paths(m, n, memo=None):
    if memo is None:
        memo = {}
    if m == 0 or n == 0:
        return 1
    if (m, n) in memo:
        return memo[(m, n)]
    memo[(m, n)] = grid_paths(m - 1, n, memo) + grid_paths(m, n - 1, memo)
    return memo[(m, n)]
print(grid_paths(2, 2))

File: 3rd_part_RO_lab_algorithms_CLI/c/ex5_grid_paths.c
#include <stdio.h>
int memo[100][100] = {0};
int grid_paths(int m, int n) {
    if (m == 0 || n == 0) return 1;
    if (memo[m][n] != 0) return memo[m][n];
    memo[m][n] = grid_paths(m - 1, n) + grid_paths(m, n - 1);
    return memo[m][n];
}
int main() {
    printf("%d\n", grid_paths(2, 2));
    return 0;
}
